azkaban-common/src/main/java/azkaban/imagemgmt/rampup/ImageRampupManagerImpl.java

```diff
@@ -0,0 +1,190 @@
+   * deprecated, and latest active version will be selected from the image_verions table. 6. If
+   * there is no active version in the image_versions table, it will throw appropriate error message
+   * mentioning could not select version for the image type and the whole process would fail. 7.
+   * Follow the rampup procedure to elect a new version from the image_versions table for the ailed
+   * image type.
+   *
+   * @param imageTypes       - set of specified image types
+   * @param imageTypeRampups - contains rampup list for an image type
+   * @return Map<String, String>
+   */
+  private Map<String, String> processAndGetVersionForImageTypes(Set<String> imageTypes,
+      Map<String, List<ImageRampup>> imageTypeRampups) {
+    Set<String> imageTypeSet = imageTypeRampups.keySet();
+    log.info("Found active rampup for the image types {} ", imageTypeSet);
+    Iterator<String> iterator = imageTypeSet.iterator();
+    Map<String, String> imageTypeVersionMap = new TreeMap<>();
+    while (iterator.hasNext()) {
+      String imageTypeName = iterator.next();
+      List<ImageRampup> imageRampupList = imageTypeRampups.get(imageTypeName);
+      Collections.sort(imageRampupList, this.getRampupPercentageComparator());
+      int prevRampupPercentage = 0;
+      int nextRandom = this.getRandomNumberInRange(1, 100);
+      for (ImageRampup imageRampup : imageRampupList) {
+        int rampupPercentage = imageRampup.getRampupPercentage();
+        if (nextRandom >= prevRampupPercentage + 1
+            && nextRandom <= prevRampupPercentage + rampupPercentage) {
+          imageTypeVersionMap.put(imageTypeName, imageRampup.getImageVersion());
+          log.info("The image version {} is selected for image type {} with rampup percentage {}",
+              imageRampup.getImageVersion(), imageTypeName, rampupPercentage);
+          break;
+        }
+        prevRampupPercentage = rampupPercentage;
+      }
+    }
+
+    /**
+     * Fetching the latest active image version from image_versions table for the remaining image
+     * types for which there is no active rampup plan or the versions are marrked as
+     * unstable/deprecated in the active plan.
+     */
+    Set<String> rampedImageTypes = imageTypeRampups.keySet();
+    Set<String> remainingImageTypes = new TreeSet<>(imageTypes);
+    remainingImageTypes.removeAll(rampedImageTypes);
+    if (!rampedImageTypes.isEmpty()) {
+      List<ImageVersion> imageVersions =
+          imageVersionDao.getActiveVersionByImageTypes(remainingImageTypes);
+      if (imageVersions != null && !imageVersions.isEmpty()) {
+        for (ImageVersion imageVersion : imageVersions) {
+          imageTypeVersionMap.put(imageVersion.getName(), imageVersion.getVersion());
+        }
+      }
+    }
+
+    // For the leftover image types throw exception with appropriate error message.
+    remainingImageTypes = new TreeSet<>(imageTypes);
```

Comment:
Any thought about keeping rampup plan in cache?

Reply:
There should be only one active rampup plan for each image type and we need to make sure we always pull the active plan and the associated rampup. It would be good to start with DB initially and in future if we see a need we can think of cache. As of now I don't think cache is required.

Reply:
Make sense. Thanks.